test/extensions/filters/http/on_demand/on_demand_integration_test.cc

```diff
@@ -761,5 +761,231 @@ TEST_P(OnDemandVhdsIntegrationTest, AttemptAddingDuplicateDomainNames) {
+
+  // Respond with a redirect to the same host but different path
+  Http::TestResponseHeaderMapImpl redirect_response{
+      {":status", "302"},
+      {"content-length", "0"},
+      {"location", "http://vhost.redirect/redirected/path"},
+      {"test-header", "redirect-value"}};
+  upstream_request_->encodeHeaders(redirect_response, true);
+
+  // Wait for the second upstream request (after redirect)
+  waitForNextUpstreamRequest(1);
+  EXPECT_EQ(request_body, upstream_request_->body().toString());
+  EXPECT_EQ("vhost.redirect", upstream_request_->headers().getHostValue());
+  EXPECT_EQ("/redirected/path", upstream_request_->headers().getPathValue());
+  ASSERT(upstream_request_->headers().EnvoyOriginalUrl() != nullptr);
+  EXPECT_EQ("http://vhost.redirect/", upstream_request_->headers().getEnvoyOriginalUrlValue());
+
+  // Send final response
+  upstream_request_->encodeHeaders(default_response_headers_, true);
+
+  // Verify the response
+  response->waitForHeaders();
+  ASSERT_TRUE(response->waitForEndStream());
+  ASSERT_TRUE(response->complete());
+  EXPECT_EQ("200", response->headers().getStatusValue());
+
+  // Verify internal redirect succeeded
+  EXPECT_EQ(1,
+            test_server_->counter("cluster.my_service.upstream_internal_redirect_succeeded_total")
+                ->value());
+  // 302 was never returned downstream
+  EXPECT_EQ(0, test_server_->counter("http.config_test.downstream_rq_3xx")->value());
+  // We expect 2 total 2xx responses: one from initial test + one from our VHDS redirect test
+  EXPECT_EQ(2, test_server_->counter("http.config_test.downstream_rq_2xx")->value());
+
+  cleanupUpstreamAndDownstream();
+}
+
 } // namespace
+
+namespace Extensions {
+namespace HttpFilters {
+namespace OnDemand {
+
+using OnDemandIntegrationTest = VhdsIntegrationTest;
+
+INSTANTIATE_TEST_SUITE_P(IpVersionsAndGrpcTypes, OnDemandIntegrationTest,
+                         DELTA_SOTW_GRPC_CLIENT_INTEGRATION_PARAMS);
+
+// Integration test specifically for the GitHub issue #17891 fix:
+// Verify that VHDS requests with body don't result in 404 responses
+TEST_P(OnDemandIntegrationTest, VhdsWithRequestBodySuccess) {
+  autonomous_upstream_ = true;
+  initialize();
+
+  // Send a POST request with body to trigger VHDS discovery
+  const std::string request_body = "test request body data";
+  auto response = codec_client_->makeRequestWithBody(
+      Http::TestRequestHeaderMapImpl{
+          {":method", "POST"},
+          {":path", "/test"},
+          {":scheme", "http"},
+          {":authority", fmt::format("vhds.example.com:{}", lookupPort("http"))},
+      },
+      request_body);
+
+  // Wait for the response - this should NOT be a 404
+  ASSERT_TRUE(response->waitForEndStream());
+  EXPECT_TRUE(response->complete());
+
+  // Before the fix, this would be "404" due to route being nullptr
+  // After the fix, this should be "200" from the upstream
+  EXPECT_EQ("200", response->headers().getStatusValue());
```

Comment:
Can you please verify if this test is correct? I run this test with "old" code and did not get 404, but timeout.